src/phoenix/trace/llama_index/callback.py

```diff
@@ -265,17 +292,68 @@ def get_spans(self) -> Iterator[Span]:
         return self._tracer.get_spans()
 
 
-def _add_to_tracer(
-    event_id_to_event_data: Dict[CBEventID, CBEventData],
-    trace_map: Dict[CBEventID, List[CBEventID]],
+def _process_exceptions(
+    event_id_to_event_data: EventMap, trace_map: TraceMap
+) -> Tuple[EventMap, TraceMap]:
+    """
+    Processes and removes exception events from the trace map. Returns copies of the inputs in which
+    the exception data has been added to the appropriate event payload and the exception events have
+    been removed.
+
+    As context, exception events are siblings and not children of the event that caused the
+    exception. There is a bug within LlamaIndex causing duplicate exception events to be emitted for
+    a single exception. This function uses a right-to-left BFS to collect exception events and then
```

Comment:
See my other comment about not having to construct a brand new tree. In that case BFS and DFS are equivalent. Since our final traversal is DFS, we might as well keep doing it the same way here without having to introduce new concepts and new data structures (i.e. `deque` and `reveresd`). 

Lastly, since the traversal here can be made identical to the final traversal, what is the benefit for duplicating the traversal just for the sake of exception attribution when the attribution itself can be done in exactly the same manner in the final traversal?

Reply:
addressed in [b8dc1e4](https://github.com/Arize-ai/phoenix/pull/1449/commits/b8dc1e423ee7b8d596d949963afc14d6f9fa2a6a)